README
o-----------------------------------------------------------------------------o
|                                                                             |
|                                 c a l 3 d                                   |
|                                                                             |
|                              Version 0.8devel                               |
|                              (in development)                               |
|                                                                             |
|            Copyright (C) 2001, 2002 Bruno 'Beosil' Heidelberger             |
|                                                                             |
o-----------------------------------------------------------------------------o

o-----------------------------------------------------------------------------o
| Table of Contents                                                           |
o-----------------------------------------------------------------------------o

  1 License
  2 Expectations
  3 What is 'cal3d'?
  4 Features
  5 Requirements
  6 Website
  7 Author


o-----------------------------------------------------------------------------o
| 1 License                                                                   |
o-----------------------------------------------------------------------------o

  This library is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as published by the
  Free Software Foundation; either version 2.1 of the License, or (at your
  option) any later version.

  This library is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
  for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this library; if not, write to the Free Software Foundation,
  Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


o-----------------------------------------------------------------------------o
| 2 Expectations                                                              |
o-----------------------------------------------------------------------------o

  The Cal3D library is licensed under the LGPL (GNU Lesser General Public
  License) which, among other things, allows you to link proprietary
  applications with it. The Cal3D team strongly believes in the Free Software
  movement and provides this library without any additional restrictions.
  However, to make this a "fair deal" and to support the future development of
  the library, we ask you to follow these points when using the Cal3D library
  in your projects:

    - Seriously consider releasing you project under a free license yourself.
    - Read and understand the LGPL and make sure that your usage of Cal3D
      complies with it.
    - Include a prominent mention about the usage of the Cal3D library in
      your project.
    - Send a short message to us when you release your work.
    - Provide feedback to the Cal3D project to help us enhance it.
    - If you make millions with your project, donate some bucks to the Cal3D
      project to fund the future development =)

  Furthermore, here are a few other possibilities to help the Cal3D project:

    - The Cal3D is always in need of models, animations, textures and demos.
      So, if you have something to spare, send it along!
    - Sponsoring and donations can definitively help the project, just
      contact us.
    - Have you heard of a new animation technology? Is there a freshly
      written, earth-shaking paper about inverse-kinematics on your desk?
      Fell free to send it to the Cal3D project, so we can put it into the
      library!

  The Cal3D team thanks you for your support and wishes you success with all
  your projects!


o-----------------------------------------------------------------------------o
| 3 What is 'cal3d'?                                                          |
o-----------------------------------------------------------------------------o

  Cal3D is a skeletal based character animation library. It is platform-
  independent and not bound to a specific graphic API. Originally designed to
  be used in a 3d client for the Worldforge project ( www.worldforge.org )
  it evolved into a stand-alone product that can be used in many different
  projects.


o-----------------------------------------------------------------------------o
| 4 Features                                                                  |
o-----------------------------------------------------------------------------o

  - skeletal based animation setup
  - seamless blending of multiple animation tracks
  - attachment and detachment of one or more skinned meshes at runtime
  - transformation of vertices and normals according to the weighted influence
    of one or more bones
  - flexible material handling
  - support for multiple texture mapping channels
  - progressive meshes for LOD (level-of-detail)
  - experimental spring-system module for cloth and hair animation
  - unified exporter plugin framework (3D Studio Max, MilkShape 3D and others
    following)
  - highly portable and modular design (only depends on C++ and STL)
  - simple and clean API (both in C++ and "pure" C)
  - did I mention that it is free? (see the LGPL license)


o-----------------------------------------------------------------------------o
| 5 Requirements                                                              |
o-----------------------------------------------------------------------------o

  There are no special requirements to compile/use cal3d. A C++ compiler with
  (decent) STL support should be enough.


o-----------------------------------------------------------------------------o
| 6 Website                                                                   |
o-----------------------------------------------------------------------------o

  The official website of cal3d can be found at: http://cal3d.sourceforge.net


o-----------------------------------------------------------------------------o
| 7 Author                                                                    |
o-----------------------------------------------------------------------------o

  This project was originally started by Bruno 'Beosil' Heidelberger. You can
  find a list of all people contributing to cal3d in the AUTHORS file.


o-----------------------------------------------------------------------------o


trunk/ChangeLog
o-----------------------------------------------------------------------------o
|                             C h a n g e L o g                               |
o-----------------------------------------------------------------------------o


o-----------------------------------------------------------------------------o
| Version 0.8 (in development)                                                |
o-----------------------------------------------------------------------------o

  - Completely revamped the automake/autoconf setup.
  - Moved the 'cal3d.doxygen' file into the 'docs' subdirectory where it
    actually belongs.
  - Added a pure C wrapper to Cal3D. This should allow non C++ applications to
    link with it now. (Thanks to Tomek Grudzinski!)
  - Changed the error codes in CalError and the type and state codes in
    CalAnimation to an enum structure.
  - Changed the interpolation calculations in CalCoretrack::getState() so that
    the motion wraps are handled like in the 3d modellers.
  - Tagged some forgotten friend functions of the CalVector and CalQuaternion
    classes as exported.


o-----------------------------------------------------------------------------o
| Version 0.7 (5. January 2002)                                               |
o-----------------------------------------------------------------------------o

  - The header files of Cal3D are now installed to '/usr/local/include/cal3d'
    when executing the 'make install' command. (Thanks to Brandon Ehle!)
  - The skeleton pose is now set to the initial state when no active animation
    exists. (Thanks to Brandon Ehle!)
  - Enhanced the CalVector and CalQuaternion class with a great number of
    operators. (Thanks to Wolfgang Scheicher!)
  - CalVector is now used as type of the vertex position and normal data in
    the CalCoreSubmesh and CalSubmesh class.
  - Revamped the animation pipeline: Added a CalPhysique helper class and a
    CalSpringSystem class.
  - Experimental cloth animation system implemented in the CalSpringSystem
    class. (Thanks to Wolfgang Scheicher!)
  - Changed the whole bone influence system to a localized bone transformation
    approach. This was needed for the planned cloth animation system and for a
    better vertex shader support.
  - Moved the 'weight' member variable and its accessor from the derived
    classes to the CalAnimation class.
  - Removed the unused 'duration' argument from the CalLoader::loadCoreTrack()
    and CalSaver::saveCoreTrack() functions.
  - Moved the setMaterialSet() function from the CalModel class to the
    CalRenderer class.
  - Relicensed the library under the LGPL (GNU Lesser General Public License).
  - CalCoreSkeleton::addCoreBone() inserts root core bones automatically into
    the root core bone list now. The function addRootCoreBoneId() was removed.


o-----------------------------------------------------------------------------o
| Version 0.6 (4. August 2001)                                                |
o-----------------------------------------------------------------------------o

  - Fixed the bone/vertex/normal transformation pipeline. There was a bug in
    the library and another in the exporter. Together they worked pretty much
    correct, but there were those weird normals in some of the models.
  - Added a user-data member variable and accessors to the CalCoreBone,
    CalCoreMaterial, CalCoreModel and CalModel classes.
  - Finished the LOD handling in the CalRenderer, CalMesh and CalCoreMesh
    classes.
  - Implemented the whole material system with material-sets and -threads.
  - I decided to implement a material class (CalCoreMaterial) and a new file
    type for it (.crf). This should lead to a very flexible material handling.
    (such as swapping materials on the fly on submeshes etc.).
  - Texture coordinates are stored differently in the CalCoreSubmesh class to
    optimize access from the CalRenderer interface.
  - A new class CalRenderer was added that serves as a clean interface for
    the external rendering loop.
  - Fixed all *.cpp files to include config.h if HAVE_CONFIG_H is set.
  - Converted CalModel::getMeshVertices() and CalModel::getMeshNormals() to
    CalModel::getSubmeshVertices() and CalModel::getSubmeshNormals().
  - Created a little model viewer named 'MiniViewer' that can be run from the
    command-line.
  - Complete redesign of the mesh handling. It is now much cleaner and ready
    for LOD (level of detail).
  - Changed the fileformats to include a version info field. Furthermore, the
    magic file cookies were modified to reflect the incompatibility to the
    old fileformat.

src/cal3d/cal3d.h
//****************************************************************************//
// cal3d.h                                                                    //
// Copyright (C) 2001, 2002 Bruno 'Beosil' Heidelberger                       //
//****************************************************************************//
// This library is free software; you can redistribute it and/or modify it    //
// under the terms of the GNU Lesser General Public License as published by   //
// the Free Software Foundation; either version 2.1 of the License, or (at    //
// your option) any later version.                                            //
//****************************************************************************//

#ifndef CAL_CAL3D_H
#define CAL_CAL3D_H

//****************************************************************************//
// Includes                                                                   //
//****************************************************************************//

#include "animation.h"
#include "animation_action.h"
#include "animation_cycle.h"
#include "bone.h"
#include "coreanimation.h"
#include "corebone.h"
#include "corekeyframe.h"
#include "corematerial.h"
#include "coremesh.h"
#include "coremodel.h"
#include "coreskeleton.h"
#include "coresubmesh.h"
#include "coretrack.h"
#include "error.h"
#include "loader.h"
#include "mesh.h"
#include "mixer.h"
#include "model.h"
#include "skiner.h"
#include "quaternion.h"
#include "renderer.h"
#include "saver.h"
#include "skeleton.h"
#include "springsystem.h"
#include "submesh.h"
#include "vector.h"

#endif

 /*****************************************************************************/
/** \mainpage Cal3D API Reference
  * <center>
  *   \image html ../../shared/cal3d.gif
  *   <p>
  *   &copy; 2001, 2002 Bruno Heidelberger
  *   <p>
  *     Permission is granted to copy, distribute and/or modify this document
  *     under the terms of the GNU Free Documentation License, Version 1.1 or
  *     any later version published by the Free Software Foundation;
  *     A copy of the license is included in the section entitled
  *     \link license "GNU Free Documentation License" \endlink .
  * </center>
  *****************************************************************************/

 /*****************************************************************************/
/** \page license GNU Free Documentation License
  * Version 1.1, March 2000
  * <p>
  * Copyright (C) 2000  Free Software Foundation, Inc.<BR>
  * 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA<BR>
  * Everyone is permitted to copy and distribute verbatim copies<BR>
  * of this license document, but changing it is not allowed.<BR>
  * <p>
  * <strong>0. PREAMBLE</strong>
  * <p>
  * The purpose of this License is to make a manual, textbook, or other
  * written document "free" in the sense of freedom: to assure everyone
  * the effective freedom to copy and redistribute it, with or without
  * modifying it, either commercially or noncommercially.  Secondarily,
  * this License preserves for the author and publisher a way to get
  * credit for their work, while not being considered responsible for
  * modifications made by others.
  * <p>
  * This License is a kind of "copyleft", which means that derivative
  * works of the document must themselves be free in the same sense.  It
  * complements the GNU General Public License, which is a copyleft
  * license designed for free software.
  * <p>
  * We have designed this License in order to use it for manuals for free
  * software, because free software needs free documentation: a free
  * program should come with manuals providing the same freedoms that the
  * software does.  But this License is not limited to software manuals;
  * it can be used for any textual work, regardless of subject matter or
  * whether it is published as a printed book.  We recommend this License
  * principally for works whose purpose is instruction or reference.
  * <p>
  *
  * <strong>1. APPLICABILITY AND DEFINITIONS</strong>
  * <p>
  * This License applies to any manual or other work that contains a
  * notice placed by the copyright holder saying it can be distributed
  * under the terms of this License.  The "Document", below, refers to any
  * such manual or work.  Any member of the public is a licensee, and is
  * addressed as "you".
  * <p>
  * A "Modified Version" of the Document means any work containing the
  * Document or a portion of it, either copied verbatim, or with
  * modifications and/or translated into another language.
  * <p>
  * A "Secondary Section" is a named appendix or a front-matter section of
  * the Document that deals exclusively with the relationship of the
  * publishers or authors of the Document to the Document's overall subject
  * (or to related matters) and contains nothing that could fall directly
  * within that overall subject.  (For example, if the Document is in part a
  * textbook of mathematics, a Secondary Section may not explain any
  * mathematics.)  The relationship could be a matter of historical
  * connection with the subject or with related matters, or of legal,
  * commercial, philosophical, ethical or political position regarding
  * them.
  * <p>
  * The "Invariant Sections" are certain Secondary Sections whose titles
  * are designated, as being those of Invariant Sections, in the notice
  * that says that the Document is released under this License.
  * <p>
  * The "Cover Texts" are certain short passages of text that are listed,
  * as Front-Cover Texts or Back-Cover Texts, in the notice that says that
  * the Document is released under this License.
  * <p>
  * A "Transparent" copy of the Document means a machine-readable copy,
  * represented in a format whose specification is available to the
  * general public, whose contents can be viewed and edited directly and
  * straightforwardly with generic text editors or (for images composed of
  * pixels) generic paint programs or (for drawings) some widely available
  * drawing editor, and that is suitable for input to text formatters or
  * for automatic translation to a variety of formats suitable for input
  * to text formatters.  A copy made in an otherwise Transparent file
  * format whose markup has been designed to thwart or discourage
  * subsequent modification by readers is not Transparent.  A copy that is
  * not "Transparent" is called "Opaque".
  * <p>
  * Examples of suitable formats for Transparent copies include plain
  * ASCII without markup, Texinfo input format, LaTeX input format, SGML
  * or XML using a publicly available DTD, and standard-conforming simple
  * HTML designed for human modification.  Opaque formats include
  * PostScript, PDF, proprietary formats that can be read and edited only
  * by proprietary word processors, SGML or XML for which the DTD and/or
  * processing tools are not generally available, and the
  * machine-generated HTML produced by some word processors for output
  * purposes only.
  * <p>
  * The "Title Page" means, for a printed book, the title page itself,
  * plus such following pages as are needed to hold, legibly, the material
  * this License requires to appear in the title page.  For works in
  * formats which do not have any title page as such, "Title Page" means
  * the text near the most prominent appearance of the work's title,
  * preceding the beginning of the body of the text.
  * <p>
  *
  * <strong>2. VERBATIM COPYING</strong>
  * <p>
  * You may copy and distribute the Document in any medium, either
  * commercially or noncommercially, provided that this License, the
  * copyright notices, and the license notice saying this License applies
  * to the Document are reproduced in all copies, and that you add no other
  * conditions whatsoever to those of this License.  You may not use
  * technical measures to obstruct or control the reading or further
  * copying of the copies you make or distribute.  However, you may accept
  * compensation in exchange for copies.  If you distribute a large enough
  * number of copies you must also follow the conditions in section 3.
  * <p>
  * You may also lend copies, under the same conditions stated above, and
  * you may publicly display copies.
  * <p>
  *
  * <strong>3. COPYING IN QUANTITY</strong>
  * <p>
  * If you publish printed copies of the Document numbering more than 100,
  * and the Document's license notice requires Cover Texts, you must enclose
  * the copies in covers that carry, clearly and legibly, all these Cover
  * Texts: Front-Cover Texts on the front cover, and Back-Cover Texts on
  * the back cover.  Both covers must also clearly and legibly identify
  * you as the publisher of these copies.  The front cover must present
  * the full title with all words of the title equally prominent and
  * visible.  You may add other material on the covers in addition.
  * Copying with changes limited to the covers, as long as they preserve
  * the title of the Document and satisfy these conditions, can be treated
  * as verbatim copying in other respects.
  * <p>
  * If the required texts for either cover are too voluminous to fit
  * legibly, you should put the first ones listed (as many as fit
  * reasonably) on the actual cover, and continue the rest onto adjacent
  * pages.
  * <p>
  * If you publish or distribute Opaque copies of the Document numbering
  * more than 100, you must either include a machine-readable Transparent
  * copy along with each Opaque copy, or state in or with each Opaque copy
  * a publicly-accessible computer-network location containing a complete
  * Transparent copy of the Document, free of added material, which the
  * general network-using public has access to download anonymously at no
  * charge using public-standard network protocols.  If you use the latter
  * option, you must take reasonably prudent steps, when you begin
  * distribution of Opaque copies in quantity, to ensure that this
  * Transparent copy will remain thus accessible at the stated location
  * until at least one year after the last time you distribute an Opaque
  * copy (directly or through your agents or retailers) of that edition to
  * the public.
  * <p>
  * It is requested, but not required, that you contact the authors of the
  * Document well before redistributing any large number of copies, to give
  * them a chance to provide you with an updated version of the Document.
  * <p>
  *
  * <strong>4. MODIFICATIONS</strong>
  * <p>
  * You may copy and distribute a Modified Version of the Document under
  * the conditions of sections 2 and 3 above, provided that you release
  * the Modified Version under precisely this License, with the Modified
  * Version filling the role of the Document, thus licensing distribution
  * and modification of the Modified Version to whoever possesses a copy
  * of it.  In addition, you must do these things in the Modified Version:
  * <p>
  * <ul>
  * <li><strong>A.</strong> Use in the Title Page (and on the covers, if any) a title distinct
  *    from that of the Document, and from those of previous versions
  *    (which should, if there were any, be listed in the History section
  *    of the Document).  You may use the same title as a previous version
  *    if the original publisher of that version gives permission.
  * <li><strong>B.</strong> List on the Title Page, as authors, one or more persons or entities
  *    responsible for authorship of the modifications in the Modified
  *    Version, together with at least five of the principal authors of the
  *    Document (all of its principal authors, if it has less than five).
  * <li><strong>C.</strong> State on the Title page the name of the publisher of the
  *    Modified Version, as the publisher.
  * <li><strong>D.</strong> Preserve all the copyright notices of the Document.
  * <li><strong>E.</strong> Add an appropriate copyright notice for your modifications
  *    adjacent to the other copyright notices.
  * <li><strong>F.</strong> Include, immediately after the copyright notices, a license notice
  *    giving the public permission to use the Modified Version under the
  *    terms of this License, in the form shown in the Addendum below.
  * <li><strong>G.</strong> Preserve in that license notice the full lists of Invariant Sections
  *    and required Cover Texts given in the Document's license notice.
  * <li><strong>H.</strong> Include an unaltered copy of this License.
  * <li><strong>I.</strong> Preserve the section entitled "History", and its title, and add to
  *    it an item stating at least the title, year, new authors, and
  *    publisher of the Modified Version as given on the Title Page.  If
  *    there is no section entitled "History" in the Document, create one
  *    stating the title, year, authors, and publisher of the Document as
  *    given on its Title Page, then add an item describing the Modified
  *    Version as stated in the previous sentence.
  * <li><strong>J.</strong> Preserve the network location, if any, given in the Document for
  *    public access to a Transparent copy of the Document, and likewise
  *    the network locations given in the Document for previous versions
  *    it was based on.  These may be placed in the "History" section.
  *    You may omit a network location for a work that was published at
  *    least four years before the Document itself, or if the original
  *    publisher of the version it refers to gives permission.
  * <li><strong>K.</strong> In any section entitled "Acknowledgements" or "Dedications",
  *    preserve the section's title, and preserve in the section all the
  *    substance and tone of each of the contributor acknowledgements
  *    and/or dedications given therein.
  * <li><strong>L.</strong> Preserve all the Invariant Sections of the Document,
  *    unaltered in their text and in their titles.  Section numbers
  *    or the equivalent are not considered part of the section titles.
  * <li><strong>M.</strong> Delete any section entitled "Endorsements".  Such a section
  *    may not be included in the Modified Version.
  * <li><strong>N.</strong> Do not retitle any existing section as "Endorsements"
  *    or to conflict in title with any Invariant Section.
  * </ul>
  * <p>
  * If the Modified Version includes new front-matter sections or
  * appendices that qualify as Secondary Sections and contain no material
  * copied from the Document, you may at your option designate some or all
  * of these sections as invariant.  To do this, add their titles to the
  * list of Invariant Sections in the Modified Version's license notice.
  * These titles must be distinct from any other section titles.
  * <p>
  * You may add a section entitled "Endorsements", provided it contains
  * nothing but endorsements of your Modified Version by various
  * parties--for example, statements of peer review or that the text has
  * been approved by an organization as the authoritative definition of a
  * standard.
  * <p>
  * You may add a passage of up to five words as a Front-Cover Text, and a
  * passage of up to 25 words as a Back-Cover Text, to the end of the list
  * of Cover Texts in the Modified Version.  Only one passage of
  * Front-Cover Text and one of Back-Cover Text may be added by (or
  * through arrangements made by) any one entity.  If the Document already
  * includes a cover text for the same cover, previously added by you or
  * by arrangement made by the same entity you are acting on behalf of,
  * you may not add another; but you may replace the old one, on explicit
  * permission from the previous publisher that added the old one.
  * <p>
  * The author(s) and publisher(s) of the Document do not by this License
  * give permission to use their names for publicity for or to assert or
  * imply endorsement of any Modified Version.
  * <p>
  *
  * <strong>5. COMBINING DOCUMENTS</strong>
  * <p>
  * You may combine the Document with other documents released under this
  * License, under the terms defined in section 4 above for modified
  * versions, provided that you include in the combination all of the
  * Invariant Sections of all of the original documents, unmodified, and
  * list them all as Invariant Sections of your combined work in its
  * license notice.
  * <p>
  * The combined work need only contain one copy of this License, and
  * multiple identical Invariant Sections may be replaced with a single
  * copy.  If there are multiple Invariant Sections with the same name but
  * different contents, make the title of each such section unique by
  * adding at the end of it, in parentheses, the name of the original
  * author or publisher of that section if known, or else a unique number.
  * Make the same adjustment to the section titles in the list of
  * Invariant Sections in the license notice of the combined work.
  * <p>
  * In the combination, you must combine any sections entitled "History"
  * in the various original documents, forming one section entitled
  * "History"; likewise combine any sections entitled "Acknowledgements",
  * and any sections entitled "Dedications".  You must delete all sections
  * entitled "Endorsements."
  * <p>
  *
  * <strong>6. COLLECTIONS OF DOCUMENTS</strong>
  * <p>
  * You may make a collection consisting of the Document and other documents
  * released under this License, and replace the individual copies of this
  * License in the various documents with a single copy that is included in
  * the collection, provided that you follow the rules of this License for
  * verbatim copying of each of the documents in all other respects.
  * <p>
  * You may extract a single document from such a collection, and distribute
  * it individually under this License, provided you insert a copy of this
  * License into the extracted document, and follow this License in all
  * other respects regarding verbatim copying of that document.
  * <p>
  *
  * <strong>7. AGGREGATION WITH INDEPENDENT WORKS</strong>
  * <p>
  * A compilation of the Document or its derivatives with other separate
  * and independent documents or works, in or on a volume of a storage or
  * distribution medium, does not as a whole count as a Modified Version
  * of the Document, provided no compilation copyright is claimed for the
  * compilation.  Such a compilation is called an "aggregate", and this
  * License does not apply to the other self-contained works thus compiled
  * with the Document, on account of their being thus compiled, if they
  * are not themselves derivative works of the Document.
  * <p>
  * If the Cover Text requirement of section 3 is applicable to these
  * copies of the Document, then if the Document is less than one quarter
  * of the entire aggregate, the Document's Cover Texts may be placed on
  * covers that surround only the Document within the aggregate.
  * Otherwise they must appear on covers around the whole aggregate.
  * <p>
  *
  * <strong>8. TRANSLATION</strong>
  * <p>
  * Translation is considered a kind of modification, so you may
  * distribute translations of the Document under the terms of section 4.
  * Replacing Invariant Sections with translations requires special
  * permission from their copyright holders, but you may include
  * translations of some or all Invariant Sections in addition to the
  * original versions of these Invariant Sections.  You may include a
  * translation of this License provided that you also include the
  * original English version of this License.  In case of a disagreement
  * between the translation and the original English version of this
  * License, the original English version will prevail.
  * <p>
  *
  * <strong>9. TERMINATION</strong>
  * <p>
  * You may not copy, modify, sublicense, or distribute the Document except
  * as expressly provided for under this License.  Any other attempt to
  * copy, modify, sublicense or distribute the Document is void, and will
  * automatically terminate your rights under this License.  However,
  * parties who have received copies, or rights, from you under this
  * License will not have their licenses terminated so long as such
  * parties remain in full compliance.
  * <p>
  *
  * <strong>10. FUTURE REVISIONS OF THIS LICENSE</strong>
  * <p>
  * The Free Software Foundation may publish new, revised versions
  * of the GNU Free Documentation License from time to time.  Such new
  * versions will be similar in spirit to the present version, but may
  * differ in detail to address new problems or concerns.  See
  * http://www.gnu.org/copyleft/.
  * <p>
  * Each version of the License is given a distinguishing version number.
  * If the Document specifies that a particular numbered version of this
  * License "or any later version" applies to it, you have the option of
  * following the terms and conditions either of that specified version or
  * of any later version that has been published (not as a draft) by the
  * Free Software Foundation.  If the Document does not specify a version
  * number of this License, you may choose any version ever published (not
  * as a draft) by the Free Software Foundation.
  *****************************************************************************/

//****************************************************************************//
